algorithms/cuda/src/clusterization/component_connection_kernels.cu

```diff
@@ -0,0 +1,185 @@
+		       vecmem::data::vector_view<unsigned int> _num_labels,
+		       vecmem::data::jagged_vector_view< measurement > _ms_per_event,
+		       vecmem::data::jagged_vector_view< spacepoint > _sp_per_event);
+
+  void sparse_ccl_cuda(
+		  const vecmem::data::jagged_vector_view< cell >& cell_per_event,
+		  vecmem::data::jagged_vector_view< unsigned int>& label_per_event,
+		  vecmem::data::vector_view<unsigned int> num_labels){    
+    sparse_ccl_kernel<<< 1, cell_per_event.m_size >>>(cell_per_event,
+						    label_per_event,
+						    num_labels);        
+    VECMEM_CUDA_ERROR_CHECK(cudaGetLastError());
+    VECMEM_CUDA_ERROR_CHECK(cudaDeviceSynchronize());
+  }
+
+  void sp_formation_cuda(const vecmem::data::jagged_vector_view< cell >& cell_per_event,
+		    const vecmem::data::jagged_vector_view< unsigned int >& label_per_event,
+		    const vecmem::data::vector_view<unsigned int> num_labels,
+		    vecmem::data::jagged_vector_view< measurement >& ms_per_event,
+		    vecmem::data::jagged_vector_view< spacepoint>& sp_per_event){
+    sp_formation_kernel<<< 1, cell_per_event.m_size >>>(cell_per_event,
+						      label_per_event,
+						      num_labels,
+						      ms_per_event,
+						      sp_per_event);    
+    VECMEM_CUDA_ERROR_CHECK(cudaGetLastError());
+    VECMEM_CUDA_ERROR_CHECK(cudaDeviceSynchronize());
+  }  
+
+__global__
+void sparse_ccl_kernel(vecmem::data::jagged_vector_view< cell > _cell_per_event,
+		     vecmem::data::jagged_vector_view< unsigned int > _label_per_event,
+		     vecmem::data::vector_view<unsigned int> _num_labels){
+  int gid = blockDim.x * blockIdx.x + threadIdx.x;
+
+  vecmem::jagged_device_vector< cell > cell_per_event(_cell_per_event);
+  vecmem::jagged_device_vector< unsigned int > label_per_event(_label_per_event);
+
+  vecmem::device_vector< cell > cell_per_module = cell_per_event.at(gid);
+  vecmem::device_vector< unsigned int > label_per_module = label_per_event.at(gid);
+  vecmem::device_vector< unsigned int > num_labels(_num_labels);
+  
+  // run sparse_ccl
+  num_labels[gid] = sparse_ccl(cell_per_module, label_per_module);
+
+  //printf("%d", num_labels[gid]);
+  
+  return;
+}
+
+
+__global__
+void sp_formation_kernel(vecmem::data::jagged_vector_view< cell > _cell_per_event,
+		       vecmem::data::jagged_vector_view< unsigned int> _label_per_event,
+		       vecmem::data::vector_view<unsigned int> _num_labels,
+		       vecmem::data::jagged_vector_view< measurement > _ms_per_event,
+		       vecmem::data::jagged_vector_view< spacepoint > _sp_per_event){
+  int gid = blockDim.x * blockIdx.x + threadIdx.x;
+
+  vecmem::jagged_device_vector< cell > cell_per_event(_cell_per_event);
+  vecmem::jagged_device_vector< unsigned int > label_per_event(_label_per_event);  
+  vecmem::device_vector< unsigned int > num_labels(_num_labels);
+  vecmem::jagged_device_vector< measurement > ms_per_event(_ms_per_event);
+  vecmem::jagged_device_vector< spacepoint > sp_per_event(_sp_per_event);  
```

Comment:
Note for future vecmem dev: it would be nice if the API could be adjusted so that there is no need for such "redeclaration".

Reply:
Certainly. I am not sure if it is possible technically though. 
I think it is gonna be nice if we can ping attila here?

Reply:
I suppose you could make the types convertible from one to another so improve the ergonomics. This could of course cause problems in cases where you don't want that to occur.

Reply:
The types are already convertible.
  - https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/containers/device_vector.hpp#L63-L65
  - https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/containers/jagged_device_vector.hpp#L82-L85

But if you want to make use of a "vector type" you do need to declare its concrete type somewhere in your code.

In the meantime I also thought a bit more about the helper function idea that I brought up before. But that won't really work. Since there is no one-to-one association from data to "interactive" types in the vecmem code. (At the moment `vecmem::data::vector_view` can be used as an input to both `vecmem::device_vector` and `vecmem::device_array`.)